avocado/utils/asset.py

```diff
@@ -107,45 +123,44 @@ def fetch(self):
         # If we get to this point, we have to download it from a location.
         # A writable cache directory is then needed. The first available
         # writable cache directory will be used.
-        for cache_dir in self.cache_dirs:
-            cache_dir = os.path.expanduser(cache_dir)
-            self.asset_file = os.path.join(cache_dir, self.basename)
-            self.hashfile = '%s-CHECKSUM' % self.asset_file
-            if not utils_path.usable_rw_dir(cache_dir):
-                continue
-
-            # Now we have a writable cache_dir. Let's get the asset.
-            # Adding the user defined locations to the urls list:
-            if self.locations is not None:
-                for item in self.locations:
-                    urls.append(item)
-
-            for url in urls:
-                urlobj = urlparse.urlparse(url)
-                if urlobj.scheme in ['http', 'https', 'ftp']:
-                    try:
-                        if self._download(url):
-                            return self.asset_file
-                    except:
-                        exc_type, exc_value = sys.exc_info()[:2]
-                        log.error('%s: %s' % (exc_type.__name__, exc_value))
-
-                elif urlobj.scheme == 'file':
-                    # Being flexible with the urlparse result
-                    if os.path.isdir(urlobj.path):
-                        path = os.path.join(urlobj.path, self.name)
-                    else:
-                        path = urlobj.path
-
-                    try:
-                        if self._get_local_file(path):
-                            return self.asset_file
-                    except:
-                        exc_type, exc_value = sys.exc_info()[:2]
-                        log.error('%s: %s' % (exc_type.__name__, exc_value))
+        cache_dir = self._get_writable_cache_dir()
+        if cache_dir is None:
+            raise EnvironmentError("Can't find a writable cache directory.")
+
+        self.asset_file = os.path.join(cache_dir, self.basename)
+        self.hashfile = '%s-CHECKSUM' % self.asset_file
+
+        # Now we have a writable cache_dir. Let's get the asset.
+        # Adding the user defined locations to the urls list:
+        if self.locations is not None:
```

Comment:
perhaps it'd make sense to move the `urls = []` and `urls.append(self.nameobj.geturl())` closer to this block. I'd suggest first obtaining the `cache_dir` and then generating the `urls` altogether.

Reply:
I'll leave this to another round of refactors.